src/coin_test/data/dataset_generator.py

```diff
@@ -210,3 +214,396 @@ class ReturnsDatasetGenerator(StitchedChunkDatasetGenerator):
     def __init__(self, dataset: "ReturnsDatasetGenerator.DATASET_TYPE") -> None:
         """Initialize a ResultsDatasetGenerator object."""
         super().__init__(dataset, chunk_size=1)
+
+
+@dataclass
+class GarchSettings:
+    """Class for keeping track of settings to intialize a GARCH model.
+
+    Default values initialize a GARCH(1,1) model with constant mean.
+    """
+
+    mean: Literal[
+        "Constant", "Zero", "LS", "AR", "ARX", "HAR", "HARX", "constant", "zero"
+    ] = "Constant"
+    lags: int | NDArray | List[int] | None = 0
+    vol: Literal["GARCH", "ARCH", "EGARCH", "FIGARCH", "APARCH", "HARCH"] = "GARCH"
+    p: int | List[int] = 1
+    o: int = 0
+    q: int = 1
+    power: float = 2
+    dist: Literal[
+        "normal",
+        "gaussian",
+        "t",
+        "studentst",
+        "skewstudent",
+        "skewt",
+        "ged",
+        "generalized error",
+    ] = "normal"
+    hold_back: int | None = None
+    rescale: bool | None = None
+
+
+class GarchDatasetGenerator(DatasetGenerator):
+    """Synthetic Dataset Generator with GARCH.
+
+    Using Generalized Autoregressive Conditional Heteroskedasticity (GARCH) model.
+
+    Close prices are simulated with univariate GARCH model.
+    Open prices are set as previous day's close.
+    High and Low prices are randomly sampled chunks transformed into and reverted from
+    percent changes relative to min/max Open/Close of each period bar.
+    """
+
+    DATASET_TYPE = CustomDataset
+
+    def __init__(
+        self,
+        dataset: "GarchDatasetGenerator.DATASET_TYPE",
+        chunk_size: int = 1,
+        mean: Literal[
+            "Constant", "Zero", "LS", "AR", "ARX", "HAR", "HARX", "constant", "zero"
+        ] = "Constant",
+        lags: int | List[int] | NDArray | None = 0,
+        vol: Literal["GARCH", "ARCH", "EGARCH", "FIGARCH", "APARCH", "HARCH"] = "GARCH",
+        p: int | List[int] = 1,
+        o: int = 0,
+        q: int = 1,
+        power: float = 2,
+        dist: Literal[
+            "normal",
+            "gaussian",
+            "t",
+            "studentst",
+            "skewstudent",
+            "skewt",
+            "ged",
+            "generalized error",
+        ] = "normal",
+        hold_back: int | None = None,
+        rescale: bool | None = None,
+    ) -> None:
+        """Initializes a GARCHDatasetGenerator."""
+        self.dataset = dataset
+        self.start: pd.Period = dataset.df.index[0]  # type: ignore
+        self.metadata = dataset.metadata
+        self.chunk_size = chunk_size
```

Comment:
Chunk size doesn't seem to be used?

Reply:
I use it when sampling chunks for defining High and Low series. The GARCH is only used to simulate Close series (which is also Open series when shifted by 1 period)

Reply:
See the sample_series function